File: Algorithms/090.SubsetsII.java
public class Solution {
    public List<List<Integer>> subsetsWithDup(int[] num) {
        Arrays.sort(num);
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> current = new ArrayList<>();
        subsetsHelper(result, current, num, 0);
        return result;
    }
    private void subsetsHelper(List<List<Integer>> result, List<Integer> current, int[] S, int pos) {
        result.add(new ArrayList<>(current));
        for (int i = pos; i < S.length; i++) {
            if (i != pos && S[i] == S[i-1]) continue;
            current.add(S[i]);
            subsetsHelper(result, current, S, i+1);
            current.remove(current.size()-1);
        }
    }
    /*
    public List<List<Integer>> subsetsWithDup(int[] S) {
        Arrays.sort(S);        
        int n = S.length;
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        Set<List<Integer>> uniqueResult = new HashSet<List<Integer>>();
        for (int i=0; i < Math.pow(2, n); i++) {
            List<Integer> v = new ArrayList<>();
            int j = i;
            int k = 0;
            while (j != 0) {
                if ((j & 0x01) != 0) {
                    v.add(S[k]);
                }
                k ++;
                j >>= 1;
            }
            uniqueResult.add(v);
        }
        for (List<Integer> a:uniqueResult) {
            result.add(a);
        }
        return result;
    }
    */
}

File: Algorithms/090.SubsetsII.cpp
class Solution {
public:
    vector<vector<int> > subsetsWithDup(vector<int> &S) {
        sort(S.begin(), S.end());
        vector<vector<int> > results;
        vector<int> current;
        subsetsHelper(results, current, S, 0);
        return results;
    }
    void subsetsHelper(vector<vector<int> > &result, vector<int> &current,
            vector<int> &S, int pos) {
        result.push_back(current);
        for (int i = pos; i < S.size(); i++) {
            if (i != pos && S[i] == S[i-1]) continue;
            current.push_back(S[i]);
            subsetsHelper(result, current, S, i+1);
            current.pop_back();
        }
    }
    /*
    vector<vector<int> > subsetsWithDup(vector<int> &S) {
        sort(S.begin(), S.end());        
        int n = S.size();
        vector<vector<int> > result;
        set<vector<int> > uniqueResult;
        for (int i=0; i < pow(2, n); i++) {
            vector<int> v;
            int j = i;
            int k = 0;
            while (j != 0) {
                if (j & 0x01) {
                    v.push_back(S[k]);
                }
                k ++;
                j >>= 1;
            }
            uniqueResult.insert(v);
        }
        for (set<vector<int> >::iterator it=uniqueResult.begin(); it != uniqueResult.end(); it++) {
            result.push_back(*it);
        }
        return result;
    }
    */
};
